### PR TITLE
feat: deliver agent-generated files to bot users + docgen skill migration

### PR DESCRIPTION
**Pull request by Dhruv-Maradiya**
## Summary

This PR delivers two related capabilities. First, **agent-generated files are now delivered to bot users** (WhatsApp, Discord, Slack, Telegram): the outbound delivery service, RabbitMQ envelope, and bot adapters were extended to carry media attachments end-to-end, with shared base adapter and text-utils handling chunking and media. Second, the **legacy `generate_document` tool, `chart_utils`, and `document_utils` were removed** in favor of a docgen subagent backed by workspace-resident skills (`create-docx`, `create-pdf`, `create-pptx`, `create-spreadsheet`) that materialize templates and build scripts via the `_system` subtree. The frontend was cleaned up accordingly — the `DocumentSection` bubble, its tool renderer, and related types were deleted. New and updated unit tests cover the outbound delivery path, RabbitMQ envelope, and bot consumer/text utilities, including adversarial edge cases.

### CODE REVIEW

File: .worktreeinclude [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Copy gitignored .env files (with secrets) into every new Conductor workspace.
+# Uses .gitignore syntax. See https://conductor.build/docs/reference/files-to-copy
+apps/api/.env
+apps/web/.env
+apps/mobile/.env
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply .worktreeinclude:5]: Intentional — `.worktreeinclude` is Conductor's local-dev "files to copy" config that seeds each local workspace with the gitignored `.env` files so the dockered API/bots can run. Nothing is shipped or committed (the `.env`s stay gitignored); it only affects local Conductor workspaces on a developer's own machine. Resolving as working-as-intended.

File: apps/api/app/agents/skills/builtin/create-docx/scripts/build.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+fail() { echo "ERROR: $*" >&2; exit 1; }
+
+# --- SEAM: toolchain provisioning (idempotent) ---------------------------------
+ensure_node_pkg() {
+  local pkg="$1"
+  mkdir -p "$NODE_HOME"
+  [ -f "$NODE_HOME/package.json" ] || ( cd "$NODE_HOME" && npm init -y >/dev/null 2>&1 )
+  [ -d "$NODE_HOME/node_modules/$pkg" ] && return 0
+  ( cd "$NODE_HOME" && npm install "$pkg" >/dev/null 2>&1 ) || fail "npm install $pkg failed"
+}
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/agents/skills/builtin/create-docx/scripts/build.sh:27]: Fixed the grep-under-`set -euo pipefail` abort in a8367894b (`|| true` so a no-match grep can't swallow the error message). Leaving `npm install` unpinned for now: these `build.sh` files are explicitly a temporary toolchain-provisioning SEAM (see header) to be replaced by the JuiceFS symlink layout — version pinning will land with that work.

File: apps/api/app/agents/skills/builtin/create-docx/scripts/build.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# ESM bare imports ("docx") resolve by walking up from the SOURCE file's dir —
+# NODE_PATH applies only to CommonJS require(), not ESM. Symlink the shared
+# install next to the source (the session scratch dir is writable) so
+# `import ... from "docx"` resolves without a per-job npm install.
+ln -sfn "$NODE_HOME/node_modules" "$(dirname "$SRC")/node_modules"
+
+err="$(mktemp)"; trap 'rm -f "$err"' EXIT
+if ! node "$SRC" "$OUT" 2>"$err"; then
+  msg="$(grep -m1 -E 'Error|error' "$err" | sed 's/^[[:space:]]*//' | cut -c1-200)"
+  fail "${msg:-node failed to run $SRC (see source)}"
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/agents/skills/builtin/create-docx/scripts/build.sh:65]: Fixed in a8367894b — added `|| true` to the stderr-parsing pipeline so a no-match grep no longer aborts before `fail` under `set -euo pipefail`.

File: apps/api/app/agents/skills/builtin/create-pdf/scripts/build.sh [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  local ver="0.13.1" arch tgt
+  arch="$(uname -m)"
+  case "$arch" in
+    x86_64|amd64) tgt="x86_64-unknown-linux-musl" ;;
+    aarch64|arm64) tgt="aarch64-unknown-linux-musl" ;;
+    *) fail "0:0: unsupported arch for typst: $arch" ;;
+  esac
+  curl -fsSL "https://github.com/typst/typst/releases/download/v${ver}/typst-${tgt}.tar.xz" \
+    | tar -xJ -C "$BIN" --strip-components=1 "typst-${tgt}/typst" || fail "0:0: typst download failed"
+  chmod +x "$BIN/typst"
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/agents/skills/builtin/create-pdf/scripts/build.sh:41]: These downloads already pin exact versions (typst 0.13.1, tectonic 0.15.0) over HTTPS from the official GitHub release hosts, into an ephemeral non-root docgen sandbox. SHA-256 pinning of the multi-arch tarballs is real hardening but premature for this documented temporary SEAM (slated for replacement by the JuiceFS toolchain layout); tracking it there rather than maintaining per-arch digests here. (Also fixed the unguarded error-extraction greps in this file in a8367894b.)

File: apps/api/app/agents/skills/builtin/create-pptx/scripts/build.sh [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+fail() { echo "ERROR: $*" >&2; exit 1; }
+
+# --- SEAM: toolchain provisioning (idempotent) ---------------------------------
+ensure_node_pkg() {
+  local pkg="$1"
+  mkdir -p "$NODE_HOME"
+  [ -f "$NODE_HOME/package.json" ] || ( cd "$NODE_HOME" && npm init -y >/dev/null 2>&1 )
+  [ -d "$NODE_HOME/node_modules/$pkg" ] && return 0
+  ( cd "$NODE_HOME" && npm install "$pkg" >/dev/null 2>&1 ) || fail "npm install $pkg failed"
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/agents/skills/builtin/create-pptx/scripts/build.sh:25]: Per the repo's review guidance we don't chase theoretical races — this needs two docgen jobs hitting a fresh shared `NODE_HOME` at the same instant, and the block is otherwise idempotent. It's also part of the temporary toolchain SEAM that the JuiceFS layout will replace (moving provisioning out of the hot path entirely). Resolving.

File: apps/api/app/agents/skills/builtin/create-spreadsheet/templates/report.py [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    for col, w in widths.items():
+        ws.column_dimensions[col].width = w
+
+
+def merge_title(ws: Worksheet, span: str, text: str) -> None:
+    """Merge a cell range into one title band and style its anchor cell."""
+    ws.merge_cells(span)
+    anchor = span.split(":", maxsplit=1)[0]
+    style_title(ws[anchor], text)
+    ws.row_dimensions[ws[anchor].row].height = 26
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/agents/skills/builtin/create-spreadsheet/templates/report.py]: Fixed in a8367894b: `build_model` now returns `tuple[int, int]` and the `style_*` helpers are typed (`cell: Cell`, `value: float`). Skipping the dict→TypedDict refactor — this is a standalone few-shot template run via `python report.py`; a TypedDict adds ceremony without changing behavior, and mypy is green on it as-is.

File: apps/api/app/agents/workspace/skill_loader.py [hunk 7]
```diff
@@ -87,6 +92,28 @@ def _parse_frontmatter(raw: str) -> tuple[dict[str, str], str]:
...
+        # Skip build/junk artifacts (e.g. __pycache__/*.pyc, dotfiles) so a stray
+        # local file never gets materialized + symlinked as a skill "resource".
+        rel_parts = path.relative_to(skill_dir).parts
+        if any(part == "__pycache__" or part.startswith(".") for part in rel_parts):
+            continue
+        try:
+            content = path.read_text(encoding="utf-8")
+        except (UnicodeDecodeError, OSError):
+            continue
+        resources.append((path.relative_to(skill_dir).as_posix(), content))
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/agents/workspace/skill_loader.py:123]: Fixed in a8367894b — an unreadable resource (`OSError`) now logs a warning so broken packaging is detectable at load time. The non-UTF-8 skip stays silent since it's by design (the in-memory system-file model is text-only).

File: apps/api/app/db/rabbitmq.py [hunk 8]
```diff
@@ -129,7 +129,31 @@ async def publish_outbound(self, queue_name: str, body: bytes) -> None:
...
+                # declare_outbound_topology_on_startup eagerly declares the full
+                # topology at boot for both the API and the worker — this lazy
+                # path only runs as a post-reconnect fallback, by which point the
+                # durable queues already exist.
+                self._outbound_topology_declared = True
+                log.error(
+                    "Outbound topology redeclare rejected (divergent queue arguments); "
+                    "publishing to the existing queue. Delete or migrate it to reconcile.",
+                    error=str(e),
+                )
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/db/rabbitmq.py:156]: This exact residual is already analyzed and accepted in the in-code comment (lines 143-150): `declare_outbound_topology_on_startup` eagerly declares the full topology at boot for both the API and the worker, so by the time this lazy post-reconnect fallback runs the durable queues already exist. Flipping the flag only prevents wedging every publish on a repeating failed redeclare. Resolving as intended.

File: apps/api/app/services/chat/workspace.py [hunk 9]
```diff
@@ -102,6 +124,27 @@ async def forward_artifact_events(
...
+                        platform=bot_platform,
+                        user_id=user_id,
+                        conversation_id=conversation_id,
+                        path=path,
+                        filename=path.rsplit("/", 1)[-1],
+                        content_type=payload.get("content_type"),
+                    )
+                )
+                publish_tasks.add(pub_task)
+                pub_task.add_done_callback(publish_tasks.discard)
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/services/chat/workspace.py:147]: The bot fetches the artifact bytes lazily by path at delivery time (`downloadArtifact`), not at publish time — so it gets the current on-disk version, not a snapshot. Publishing once per path is deliberate dedup; re-publishing on every rewrite would spam the user with duplicate copies of the same file. The rare publish-after-fetch rewrite is a timing edge we don't chase per the repo's review guidance. Resolving.

File: apps/api/app/services/outbound_delivery.py [hunk 10]
```diff
@@ -84,4 +124,51 @@ async def publish_outbound_message(
...
+        log.error("publish_outbound_file: publish failed", platform=platform.value, error=str(e))
+        return False
+
+    log.info(
+        "outbound_file_published",
+        platform=platform.value,
+        queue=queue_name,
+        filename=filename,
+    )
+    return True
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/services/outbound_delivery.py:174]: The sole production caller (the `chat/workspace.py` artifact forwarder) fires `publish_outbound_file` as a detached `asyncio.create_task` and discards the result — nothing records file-delivery status today. Returning the enum here would add unused surface (the text path returns it only because its caller persists it). Happy to switch to `OutboundResult` once a caller consumes it. Resolving.

File: apps/api/app/services/storage/sessions/skills.py [hunk 11]
```diff
@@ -95,6 +95,17 @@ def materialize_skills(user_root: Path, connected_ids: set[str]) -> int:
...
+            # scripts/, …) so multi-file skills work when the shared _system
+            # subtree + symlinks are unavailable (the linker replaces these with
+            # symlinks once the subtree exists). `rel` is always contained within
+            # the skill dir (see skill_loader._load_resources), so no traversal.
+            for rel, content in skill.resources:
+                res = slug_dir / rel
+                if not matches_text(res, content):
+                    res.parent.mkdir(parents=True, exist_ok=True)
+                    res.write_text(content, encoding="utf-8")
+                    written += 1
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/services/storage/sessions/skills.py:108]: Fixed in a8367894b — `materialize_skills` now prunes fallback resource copies that left the manifest (real files only; symlinks stay owned by `link_system_files`).

File: apps/api/app/services/storage/system_workspace.py [hunk 12]
```diff
@@ -80,12 +80,12 @@ def _write() -> bool:
...
         if marker.is_file() and marker.read_text(encoding="utf-8").strip() == signature:
             return True
+        root.mkdir(parents=True, exist_ok=True)
         for f in files:
             target = root / f.rel_path
             target.parent.mkdir(parents=True, exist_ok=True)
             if not (target.is_file() and target.read_text(encoding="utf-8") == f.body):
                 target.write_text(f.body, encoding="utf-8")
-        root.mkdir(parents=True, exist_ok=True)
         marker.write_text(signature, encoding="utf-8")
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/api/app/services/storage/system_workspace.py:99]: Fixed in a8367894b — `ensure_system_subtree` now prunes files dropped from `system_files()` before stamping the signature, so removed skills/docs don't linger in the shared `_system` subtree.

File: apps/bots/telegram/src/adapter.ts [hunk 13]
```diff
@@ -383,6 +387,34 @@ export class TelegramAdapter extends BaseBotAdapter {
...
+    const mime =
+      attachment.content_type ?? contentType ?? "application/octet-stream";
+    const file = new InputFile(data, attachment.filename);
+    const caption = attachment.caption ?? undefined;
+    // sendPhoto caps around 10 MB; deliver larger images as a document so they
+    // still arrive instead of being rejected.
+    if (mime.startsWith("image/") && data.length <= TELEGRAM_PHOTO_MAX_BYTES) {
+      await this.bot.api.sendPhoto(destinationId, file, { caption });
+    } else {
+      await this.bot.api.sendDocument(destinationId, file, { caption });
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply apps/bots/telegram/src/adapter.ts]: Fixed in a8367894b — captions are truncated to the 1024-char Telegram media-caption limit before `sendPhoto`/`sendDocument`.

File: libs/shared/ts/src/bots/api/index.ts [hunk 14]
```diff
@@ -881,6 +881,42 @@ export class GaiaClient {
...
+        .split("/")
+        .map((seg) => encodeURIComponent(seg))
+        .join("/");
+      const { data, headers } = await this.client.get(
+        `/api/v1/sessions/${encodeURIComponent(conversationId)}/artifacts/${encodedPath}`,
+        {
+          responseType: "arraybuffer",
+          headers: this.userHeaders(ctx),
+          maxContentLength: 50 * 1024 * 1024,
+          maxBodyLength: 50 * 1024 * 1024,
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply libs/shared/ts/src/bots/api/index.ts]: Fixed in a8367894b — raised `downloadArtifact`'s cap to 100 MB (the largest per-platform limit, WhatsApp), so 50-100 MB artifacts reach the `OUTBOUND_FILE_LIMITS` check / graceful fallback instead of failing as transport errors.

File: libs/shared/ts/src/bots/consumer/envelope.ts [hunk 15]
```diff
@@ -13,15 +13,36 @@
...
-  /** ISO-8601 enqueue timestamp. */
-  enqueued_at: z.string(),
+/**
+ * A file the bot should deliver. Bytes are NOT inlined — the bot fetches the
+ * artifact from the backend (bot-authenticated) and uploads it to the platform.
+ */
+export const outboundAttachmentSchema = z.object({
+  conversation_id: z.string().min(1),
+  /** Artifact path relative to the session's artifacts/ dir. */
+  path: z.string().min(1),
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply libs/shared/ts/src/bots/consumer/envelope.ts]: Fixed in a8367894b — the attachment `path` schema now rejects absolute paths and any `..` segment at the queue boundary.

File: libs/shared/ts/src/bots/consumer/outbound-consumer.ts [hunk 16]
```diff
@@ -177,19 +187,63 @@ export class OutboundConsumer {
...
+        );
+        // Never requeue a file: deliverFile fetches the bytes AND uploads +
+        // sends them, and a failure can surface AFTER the platform already
+        // accepted the upload (e.g. a timeout reading the response). We can't
+        // tell a pre-send fetch failure from a post-send one, so requeueing
+        // risks delivering the file to the user twice. Dead-letter it instead —
+        // the envelope is preserved in the DLQ for inspection/manual replay.
+        this.settle(channel, () => channel.nack(msg, false, false));
+      }
+      return;
```
[the remark this answers is not in the document]
Comment (Dhruv-Maradiya) [reply libs/shared/ts/src/bots/consumer/outbound-consumer.ts:210]: The producer never co-sends both: `publish_outbound_file` enqueues an attachment-only envelope and `publish_outbound` enqueues text-only — they're separate messages. A file's companion note travels as the attachment `caption`, not `env.text`, so there's no text to drop on this branch. The schema allows both only to keep the union open. Resolving.
